onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc

```diff
@@ -24,10 +26,16 @@ namespace onnxruntime {
 namespace nnapi {
 
 ModelBuilder::ModelBuilder(const GraphViewer& graph_viewer)
-    : nnapi_(NnApiImplementation()), graph_viewer_(graph_viewer), shaper_{graph_viewer} {}
-
-int32_t ModelBuilder::GetNNAPIFeatureLevel() const {
-  return nnapi_ ? static_cast<int32_t>(nnapi_->nnapi_runtime_feature_level) : 0;
+    : nnapi_(NnApiImplementation()), graph_viewer_(graph_viewer), shaper_{graph_viewer} {
+  if (auto st = GetTargetDevices(nnapi_, target_device_option_, nnapi_target_devices_, nnapi_target_devices_detail_);
+      !st.IsOK()) {
+    LOGS_DEFAULT(WARNING) << "GetTargetDevices failed with reason:" << st.ErrorMessage()
```

Comment:
Seems like there are multiple potential states and the behavior might need to be clarified.

If feature level is < 3 we return no devices.
Otherwise we populate devices and put the CPU device last (not sure what the reason for that is), but if there's an error while populating devices we return the error and whatever devices were successfully added up until that point.

If we want to proceed when there's an error shouldn't we populate all devices we can and ensure the CPU device is valid?

But if we don't expect an error simply failing overall might be a better approach. Debugging poor performance because a device lookup failure caused a subset of devices to be added would be potentially tricky and time consuming.

Reply:
Fow some reasons I put the CPU device last. (1  Hope it's helpful to accelerate nnapi compile, just assuming nnapi internally skip the last device if it has already find one device and nnapi-reference should have the lowest priority. 2) we can easily exclude nnapi-reference for mode nnapi_disable_cpu_soft. 3) we can easily log the detail how op was assigned within NNAPI devices which is helpful for debugging.

Reply:
Just failling overall might be better. If NNAPI-reference is avaible to handle and proceed, it give a pool perf which doesn't worth a complicated solution.

Reply:
Rather than failling in model_build. Would doing all devices stuff in EP level be better?